// coc-utils/out/platform.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const process = require("process");
var OperatingSystem;
(function (OperatingSystem) {
    OperatingSystem[OperatingSystem["Unknown"] = 0] = "Unknown";
    OperatingSystem[OperatingSystem["Windows"] = 1] = "Windows";
    OperatingSystem[OperatingSystem["MacOS"] = 2] = "MacOS";
    OperatingSystem[OperatingSystem["Linux"] = 3] = "Linux";
})(OperatingSystem = exports.OperatingSystem || (exports.OperatingSystem = {}));
function getPlatformDetails() {
    let operatingSystem = OperatingSystem.Unknown;
    if (process.platform === "win32") {
        operatingSystem = OperatingSystem.Windows;
    }
    else if (process.platform === "darwin") {
        operatingSystem = OperatingSystem.MacOS;
    }
    else if (process.platform === "linux") {
        operatingSystem = OperatingSystem.Linux;
    }
    const isProcess64Bit = process.arch === "x64";
    return {
        operatingSystem,
        isOS64Bit: isProcess64Bit || process.env.hasOwnProperty("PROCESSOR_ARCHITEW6432"),
        isProcess64Bit,
    };
}
exports.getPlatformDetails = getPlatformDetails;
function getPlatformSignature() {
    const plat = getPlatformDetails();
    const os_sig = (() => {
        switch (plat.operatingSystem) {
            case OperatingSystem.Windows: return "win";
            case OperatingSystem.Linux: return "linux";
            case OperatingSystem.MacOS: return "osx";
            default: return "unknown";
        }
    })();
    const arch_sig = (() => {
        if (plat.isProcess64Bit)
            return "x64";
        else
            return "x86";
    })();
    return `${os_sig}-${arch_sig}`;
}
exports.getPlatformSignature = getPlatformSignature;
