// coc-powershell/out/client/utils.js
/*---------------------------------------------------------
 * Copyright (C) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deleteSessionFile = exports.readSessionFile = exports.waitForSessionFile = exports.writeSessionFile = exports.getDebugSessionFilePath = exports.getSessionFilePath = exports.PowerShellLanguageId = void 0;
const fs = require("fs");
const path = require("path");
const coc_utils_1 = require("coc-utils");
exports.PowerShellLanguageId = "powershell";
const sessionsFolder = path.resolve(__dirname, "..", "..", "sessions/");
const sessionFilePathPrefix = path.resolve(sessionsFolder, "PSES-VSCode-" + process.env.VSCODE_PID);
// Create the sessions path if it doesn't exist already
coc_utils_1.ensurePathExists(sessionsFolder);
function getSessionFilePath(uniqueId) {
    return `${sessionFilePathPrefix}-${uniqueId}`;
}
exports.getSessionFilePath = getSessionFilePath;
function getDebugSessionFilePath() {
    return `${sessionFilePathPrefix}-Debug`;
}
exports.getDebugSessionFilePath = getDebugSessionFilePath;
function writeSessionFile(sessionFilePath, sessionDetails) {
    coc_utils_1.ensurePathExists(sessionsFolder);
    const writeStream = fs.createWriteStream(sessionFilePath);
    writeStream.write(JSON.stringify(sessionDetails));
    writeStream.close();
}
exports.writeSessionFile = writeSessionFile;
function waitForSessionFile(sessionFilePath, callback) {
    function innerTryFunc(remainingTries, delayMilliseconds) {
        if (remainingTries === 0) {
            callback(undefined, "Timed out waiting for session file to appear.");
        }
        else if (!coc_utils_1.checkIfFileExists(sessionFilePath)) {
            // Wait a bit and try again
            setTimeout(() => { innerTryFunc(remainingTries - 1, delayMilliseconds); }, delayMilliseconds);
        }
        else {
            // Session file was found, load and return it
            callback(readSessionFile(sessionFilePath), undefined);
        }
    }
    // Try once every 2 seconds, 60 times - making two full minutes
    innerTryFunc(60, 2000);
}
exports.waitForSessionFile = waitForSessionFile;
function readSessionFile(sessionFilePath) {
    const fileContents = fs.readFileSync(sessionFilePath, "utf-8");
    return JSON.parse(fileContents);
}
exports.readSessionFile = readSessionFile;
function deleteSessionFile(sessionFilePath) {
    try {
        fs.unlinkSync(sessionFilePath);
    }
    catch (e) {
        // TODO: Be more specific about what we're catching
    }
}
exports.deleteSessionFile = deleteSessionFile;
